README
$Id: README,v 1.5 2006/12/05 20:47:56 bu Exp $

Microbrew md5sum/sha1sum/ripemd160sum
Copyright (c) 2004, Bulent Yilmaz <bu@microbrew.org>
 for Microbrew Software

LICENSE
-------

GNU GPL >= 2.  See file 'LICENSE'.

Special exemption from the GPL is given solely for the purpose of linking 
with OpenSSL which is distributed under the BSD license.  This is the only
exemption from the GPL which is given and all parts of the GPL are still
in effect including those parts which are similar to the exemption above


DESCRIPTION
-----------

Drop in replacement for md5sum from 'GNU textutils'.  I wanted to install
'md5sum' on my Solaris boxen but didn't want the trauma of installing GNU
textutils... especially when Solaris comes with 95% of those programs in
the first place.

So, here it is.  The culmination of 3 hours of coding, a pot of coffee, 3 
breaks, 2 hours of maintaining my hatred of GNU autoconf, 2 hours of 
documentation, and one Solaris package.  BTW, if you know Ulrich Drepper, 
smack him for me for making me run around in circles to figure out what his 
md5sum was doing with its command line options.


INSTALLATION
------------

Please look at the file 'INSTALL'


USAGE
-----

Efforts were taken to make these programs as compatible as possible to their
counterparts from GNU textutils.  This doesn't mean that they -are- compatible
but most people will not be able to notice the difference.

Here's a dump of 'md5sum --help'

############################################################################
$ md5sum --help
Usage: ./md5sum [<option>] <file> [<file> [...] ]
       ./md5sum [<option>] --check <file>

Note:  These options are mostly compatible with GNU md5sum
       -s, -h, and -V are not available in GNU md5sum

 -b, --binary         Read files in binary mode
 -c, --check <file>   Check MD5 sums from <file>
 -t, --text           Read files in ASCII mode

 -s, --status         Silent mode: Use exit code to determine verification

 -h, --help           Display this help message and exit
 -V, --version        Display program version and exit
############################################################################

Pretty self-explainatory.  If you must have a man page for this program, the
one for 'md5sum' from 'GNU textutils' should suit you fine.  If you -MUST- 
have one specifically for this program, feel free to write it up and send it 
to me. :^p


REPORTING BUGS
--------------

As an interlude:  It is funny how visually and conceptually similar 
'reposting bugs' and 'reporting bugs' are.  Chew on that for a bit.

And oh, yeah... You can email me the bug reports for now.


THANKS IN ORDER
---------------

* Steve Moon for a compile fix which can be found in version 0.9.4.
* Ralf S. Engelschall for a fix which can be found in version 0.9.5.


ChangeLog
$Id: ChangeLog,v 1.7 2006/12/05 20:47:56 bu Exp $

md5sha1sum 0.9.5
----------------
* Added a fix so that using 'sha1sum -c' to check sums does not give a 
  size warning.  Thanks to Ralf S. Engelschall for pointing this out to me.

md5sha1sum 0.9.4
----------------
* Changed function arguments to use 'unsigned char *' instead of 'char *'
  in do_md5sum, do_sha1sum, and do_ripemd160sum.  Thanks to Steve Moon for
  pointing this out to me.

md5sha1sum 0.9.3
----------------
* Changes so it doesn't try to get the hash of a directory.
* Added excessively verbose error output.
* Miscellaneous bug squashing.

md5sha1sum 0.9.2
----------------
* The --check <file> option can now use '-' for stdin.
* Added support for OS X in configure to find 'libssl.dylib'
* Updated the name of the program here and there to include ripemd160sum

md5sha1sum 0.9.1
----------------
* Added support for ripemd160 sums... as well as a binary.
* Updated the configure script so it will not overwrite GNU md5sum if you do
  have it installed.
* Fixed bugs in the documentation.

md5sha1sum 0.9.0 -- Initial release
-----------------------------------


config.h
/**********************************************************************
Copyright 2004, Bulent Yilmaz
for Microbrew Software

Part of Microbrew md5sum/sha1sum
GNU textutils md5sum replacement

http://www.microbrew.org/tools/md5sha1sum.html

Microbrew md5sum/sha1sum is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

MicroSieve is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with MicroSieve; if not, write to the Free Software Foundation, Inc.,
59 Temple Place, Suite 330, Boston, MA  02111-1307 USA

**********************************************************************/

#define REL_VERSION "0.9.5"
#define REL_DATE "Wed Dec  6 12:48:56 EST 2006"


macros.h
/**********************************************************************
$Id: macros.h,v 1.1 2004/08/17 15:02:48 bu Exp $

Copyright 2004, Microbrew Software

by Bulent Yilmaz <bu@microbrew.org>

Microbrew Message Center is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or (at your
option) any later version.

Microbrew Message Center is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
Public License for more details.

You should have received a copy of the GNU General Public License along with
Microbrew Message Center; if not, write to the Free Software Foundation,
Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA

**********************************************************************/

#ifndef __MACROS_H
#define __MACROS_H

/* Required to get strerror(errno) to work correctly */
#include <errno.h>

/* My favourite macro ever */
#define NNFREE(x) if((x) != NULL) free(x)

/* Some macros to make nice error output */
#define eprintf(fmt,args...) fprintf(stderr, __FILE__ ":%d: " fmt, __LINE__, args)
#define eputs(fmt) fprintf(stderr, __FILE__ ":%d: " fmt "\n", __LINE__)

#define perrprintf(fmt,args...) fprintf(stderr, __FILE__ ":%d: " fmt ": %s\n", __LINE__, args, strerror(errno))
#define perrputs(fmt) fprintf(stderr, __FILE__ ":%d: " fmt ": %s\n", __LINE__, strerror(errno))

#endif


md5sum.c
/**********************************************************************
$Id: md5sum.c,v 1.5 2006/12/05 20:47:56 bu Exp $

Copyright 2004, Bulent Yilmaz
for Microbrew Software

Part of Microbrew md5sum/sha1sum
GNU textutils md5sum replacement

http://www.microbrew.org/tools/md5sha1sum.html

Microbrew md5sum/sha1sum is free software; you can redistribute it and/or 
modify it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

MicroSieve is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with MicroSieve; if not, write to the Free Software Foundation, Inc.,
59 Temple Place, Suite 330, Boston, MA  02111-1307 USA

Special exemption from the GPL is given solely for the purpose of linking 
with OpenSSL which is distributed under the BSD license.  This is the only
exemption from the GPL which is given and all parts of the GPL are still
in effect including those parts which are similar to the exemption above.

**********************************************************************/

#include <openssl/md5.h>
#include <openssl/sha.h>
#include <openssl/ripemd.h>
#include <sys/types.h>
#include <sys/stat.h>

#include "parseopts.h"
#include "macros.h"

void do_md5sum(FILE *mfile, unsigned char *rptr) {
    char rdata[BUFSIZ];
    int rsize;
    MD5_CTX c;
    
    MD5_Init(&c);
    while(!feof(mfile)) {
        rsize = fread(rdata, sizeof(char), BUFSIZ, mfile);
        if(rsize > 0) {
            MD5_Update(&c, rdata, rsize);
        }
    }

    MD5_Final(rptr, &c);
}

void do_ripemd160sum(FILE *mfile, unsigned char *rptr) {
    char rdata[BUFSIZ];
    int rsize;
    RIPEMD160_CTX c;
    
    RIPEMD160_Init(&c);
    while(!feof(mfile)) {
        rsize = fread(rdata, sizeof(char), BUFSIZ, mfile);
        if(rsize > 0) {
            RIPEMD160_Update(&c, rdata, rsize);
        }
    }

    RIPEMD160_Final(rptr, &c);
}

void do_sha1sum(FILE *mfile, unsigned char *rptr) {
    char rdata[BUFSIZ];
    int rsize;
    SHA_CTX c;
    
    SHA1_Init(&c);
    while(!feof(mfile)) {
        rsize = fread(rdata, sizeof(char), BUFSIZ, mfile);
        if(rsize > 0) {
            SHA1_Update(&c, rdata, rsize);
        }
    }

    SHA1_Final(rptr, &c);
}

int main(int argc, char *argv[]) {
    struct opts_t *opts;
    int i,j;
    FILE* tfile;
    unsigned char sumdata[SHA_DIGEST_LENGTH];
    char psum[41];
    int fails = 0;
    int hashsize;
    struct stat stbuf;

    opts = parseopts(argc, argv);
    if(opts == NULL) {
        return -1;
    }

    if(opts->filect <= 0) {
        opts->filenames = malloc(sizeof(char *));
        if(opts->filenames == NULL) {
            perrputs("main: malloc opts->filenames");
            return -1;
        }

        opts->filenames[0] = "-";
        opts->filect = 1;
    }

    for(i = 0; i < opts->filect; i++) {
        if(!strcmp(opts->filenames[i], "-")) {
            tfile = stdin;
        } else {
            /* Stat the file to make sure its a regular file */
            if(stat(opts->filenames[i], &stbuf)) {
                fprintf(stderr, "%s: stat '%s': %s\n", argv[0], 
                    opts->filenames[i], strerror(errno));
                continue;
            }

            if(S_ISREG(stbuf.st_mode)) {
                if(opts->fmode == FMODE_BIN) 
                    tfile = fopen(opts->filenames[i], "rb");
                else
                    tfile = fopen(opts->filenames[i], "r");
            } else {
                fprintf(stderr, "%s: %s: Not a regular file\n", argv[0], 
                    opts->filenames[i]);
                continue;
            }
        }

        if(tfile == NULL) {
            perrprintf("Could not open file '%s'", opts->filenames[i]);
        } else {
            if(strstr(argv[0], "sha1sum")) {
                do_sha1sum(tfile, sumdata);
                for(j = 0; j < SHA_DIGEST_LENGTH; j++)
                    sprintf(&psum[(j*2)], "%02x", (int)sumdata[j]);
                psum[(SHA_DIGEST_LENGTH * 2)] = '\0';
            } else if(strstr(argv[0], "ripemd160sum")) {
                do_ripemd160sum(tfile, sumdata);
                for(j = 0; j < RIPEMD160_DIGEST_LENGTH; j++)
                    sprintf(&psum[(j*2)], "%02x", (int)sumdata[j]);
                psum[(RIPEMD160_DIGEST_LENGTH * 2)] = '\0';
            } else {
                do_md5sum(tfile, sumdata);
                for(j = 0; j < MD5_DIGEST_LENGTH; j++)
                    sprintf(&psum[(j*2)], "%02x", (int)sumdata[j]);
                psum[(MD5_DIGEST_LENGTH * 2)] = '\0';
            }

            if(opts->pmode == PMODE_VERIFY) {
                /* Be nice an offer a warning if hash size appears to be 
                   wrong */

                hashsize = strlen(opts->hashes[i]);
                if(strstr(argv[0], "sha1sum") && (hashsize != (SHA_DIGEST_LENGTH * 2)))
                    fputs("Warning: Hash size is wrong, maybe you want to use 'md5sum'\n", stderr);
                if(strstr(argv[0], "ripemd160sum") && (hashsize != (RIPEMD160_DIGEST_LENGTH * 2)))
                    fputs("Warning: Hash size is wrong, maybe you want to use 'md5sum'\n", stderr);
                if(strstr(argv[0], "md5sum") && hashsize != (MD5_DIGEST_LENGTH
* 2))
                    fputs("Warning: Hash size is wrong, maybe you want to use 'sha1sum' or 'ripemd160sum'\n", stderr);

                if(opts->vmode == VMODE_SILENT) {
                    if(strcmp(opts->hashes[i], psum)) fails++;
                } else {
                    printf("%s: ", opts->filenames[i]);
                    if(!strcmp(opts->hashes[i], psum)) puts("OK");
                    else {
                        puts("FAILED");
                        fails++;
                    }
                }
            } else {
                if(opts->fmode == FMODE_BIN) 
                    printf("%s *%s\n", psum, opts->filenames[i]);
                else
                    printf("%s  %s\n", psum, opts->filenames[i]);
            }
        }

        if(tfile != stdin) fclose(tfile);
    } 

    if(opts->pmode == PMODE_VERIFY) {
        if(fails > 0) {
            if(opts->vmode != VMODE_SILENT) {
                printf("%s: WARNING: %d of %d computed checksums did NOT match\n", argv[0], fails, opts->filect);
            }
            return 1;
        }
    }

    return 0;
}


parseopts.h
/**********************************************************************
$Id: parseopts.h,v 1.3 2004/08/17 15:02:48 bu Exp $

Copyright 2004, Bulent Yilmaz
for Microbrew Software

Part of Microbrew md5sum/sha1sum
GNU textutils md5sum replacement

http://www.microbrew.org/tools/md5sha1sum.html

Microbrew md5sum/sha1sum is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

MicroSieve is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with MicroSieve; if not, write to the Free Software Foundation, Inc.,
59 Temple Place, Suite 330, Boston, MA  02111-1307 USA

**********************************************************************/

#ifndef _PARSEOPTS_H
#define _PARSEOPTS_H

#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#include "config.h"
#include "macros.h"

#define FMODE_ASC 0
#define FMODE_BIN 1

#define PMODE_SUM 0
#define PMODE_VERIFY 1

#define VMODE_NORMAL 0
#define VMODE_SILENT 1

struct opts_t {
    int pmode;
    int fmode;
    int vmode;
    int filect;
    char **filenames;
    char **hashes;
};

struct opts_t *parseopts(int, char **);
#endif


parseopts.c
/**********************************************************************
$Id: parseopts.c,v 1.7 2006/12/05 20:47:56 bu Exp $

Copyright 2004, Bulent Yilmaz
for Microbrew Software

Part of Microbrew md5sum/sha1sum/ripemd160sum
GNU textutils md5sum replacement

http://www.microbrew.org/tools/md5sha1sum.html

Microbrew md5sum/sha1sum is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

MicroSieve is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with MicroSieve; if not, write to the Free Software Foundation, Inc.,
59 Temple Place, Suite 330, Boston, MA  02111-1307 USA

**********************************************************************/

#include "parseopts.h"

void showverinfo() {
    printf("Microbrew md5sum/sha1sum/ripemd160sum %s (%s)\n", 
           REL_VERSION, REL_DATE);
    printf("  Compiled %s at %s\n", __DATE__, __TIME__);
    puts("Written by Bulent Yilmaz");
    puts("");
    puts("Copyright (C) 2004,2006 Microbrew Software");
}

void showhelp(char *progname) {
    printf("Usage: %s [<option>] <file> [<file> [...] ]\n", progname);
    printf("       %s [<option>] --check <file>\n", progname);
    puts("");
    puts("Note:  These options are mostly compatible with GNU md5sum");
    puts("       -s, -h, and -V are not available in GNU md5sum");
    puts("");
    puts(" -b, --binary         Read files in binary mode");
    puts(" -c, --check <file>   Check MD5 sums from <file>");
    puts(" -t, --text           Read files in ASCII mode");
    puts("");
    puts(" -s, --status         Silent mode: Use exit code to determine verification");
    puts("");
    puts(" -h, --help           Display this help message and exit");
    puts(" -V, --version        Display program version and exit");
}

char *fixfn_strdup(char *input) {

    int spn;
    char *rstr;

    spn = strspn(input, " \t\n*");
    rstr = strdup(&input[spn]);

    return rstr;
}

int setup_check(struct opts_t *opts, char *cfile) {
    FILE *chk;
    char line[BUFSIZ];
    char *hash, *fn;
    void *holder;

    if(strcmp(cfile, "-")) {
        chk = fopen(cfile, "r");
        if(chk == NULL) {
            perrprintf("setup_check: fopen '%s'", cfile);
            return -1;
        }
    } else chk = stdin;

    while(!feof(chk)) {
#if 0
        printf("Parsing %d\n", opts->filect);
#endif

        fgets(line, BUFSIZ, chk);
        if(feof(chk)) break;

        hash = strtok(line, "\t ");
        if(hash == NULL) {
            printf("Could not parse check file '%s' (1)\n", cfile);
            if(chk != stdin) fclose(chk);
            return -1;
        } 

        fn = strtok(NULL, "\n");
        if(fn == NULL) {
            printf("Could not parse check file '%s' (2)\n", cfile);
            if(chk != stdin) fclose(chk);
            return -1;
        }

        if(strcmp(fn, "-") && strcmp(fn, "*-")) {
            holder = realloc(opts->filenames, 
                             (opts->filect + 1) * sizeof(char *));
            if(holder == NULL) {
                perrputs("setup_check: realloc(opts->filenames)");
                if(chk != stdin) fclose(chk);
                return -1;
            }
            opts->filenames = holder;

            holder = realloc(opts->hashes, (opts->filect + 1) * sizeof(char *));
            if(holder == NULL) {
                perrputs("setup_check: realloc(opts->hashes)");
                if(chk != stdin) fclose(chk);
                return -1;
            }
            opts->hashes = holder;

            opts->filenames[opts->filect] = fixfn_strdup(fn);
            opts->hashes[opts->filect++] = strdup(hash);
        }
    }

    if(chk != stdin) fclose(chk);
    opts->pmode = PMODE_VERIFY;
    return 0;
}

struct opts_t *parseopts(int argc, char **argv) {
    struct opts_t *ropts;
    int i, offs;

    ropts = malloc(sizeof(struct opts_t));
    if(ropts == NULL) {
        perrputs("parseopts: malloc ropts");
        return NULL;
    }

    ropts->fmode = FMODE_ASC; 
    ropts->pmode = PMODE_SUM;
    ropts->vmode = VMODE_NORMAL;
    ropts->filect = 0;
    ropts->filenames = NULL;
    ropts->hashes = NULL;

    for(i = 1; i < argc; i++) {
        if( (!strcmp(argv[i], "-b")) ||
            (!strcmp(argv[i], "--binary")) ) ropts->fmode = FMODE_BIN;
        else if( (!strcmp(argv[i], "-t")) ||
                 (!strcmp(argv[i], "--text")) ) ropts->fmode = FMODE_ASC;
        else if( (!strcmp(argv[i], "-s")) || (!strcmp(argv[i], "--status")) )
           ropts->vmode = VMODE_SILENT;
        else if( (!strcmp(argv[i], "-h")) || (!strcmp(argv[i], "--help")) ) {
            showhelp(argv[0]);
            free(ropts);
            return NULL;
        } else if( (!strcmp(argv[i], "-V")) || 
                   (!strcmp(argv[i], "--version")) ) {
            showverinfo();
            free(ropts);
            return NULL;
        } else if( (!strcmp(argv[i], "-c")) ||
                   (!strcmp(argv[i], "--check")) ) {
            if(argc > (i + 1)) {
                if(setup_check(ropts, argv[++i])) {
                    free(ropts);
                    return NULL;
                }
            } else {
                showhelp(argv[0]);
            }
        } else {
            /* Make sure we're not in verify mode */
            if(ropts->pmode != 0) {
                fprintf(stderr, "Error: --check <filename> cannot be used with additional files\n");
                free(ropts);
                return NULL;
            }

            /* Assume everything else is a filename */
            ropts->filect = argc - i;
            ropts->filenames = malloc(ropts->filect * sizeof(char *));
            if(ropts->filenames == NULL) {
                perrputs("parseopts: malloc ropts->filenames");
                free(ropts);
                return NULL;
            }

            offs = i;
            for(; i < argc; i++) ropts->filenames[(i - offs)] = argv[i];
        }
    }

    return ropts;
}
